crates/file_finder/src/file_finder.rs

```diff
@@ -230,88 +274,104 @@ impl Matches {
-                        match_b.cmp(match_a)
-                    }
-                    (Match::History(_, match_a), Match::Search(match_b)) => {
-                        Some(match_b).cmp(&match_a.as_ref())
-                    }
-                    (Match::Search(match_a), Match::History(_, match_b)) => {
-                        match_b.as_ref().cmp(&Some(match_a))
+        //NOTE: At this point we have an unsorted set of new history matches, an unsorted set of new search matches
+        // and a sorted set of old search matches.
+        // It is possible that the new search matches' paths contain some of the old search matches' paths.
+        // History matches' paths are unique, since store in a HashMap by path.
+        // We build a sorted Vec<Match>, eliminating duplicate search matches.
+        // Search matches with the same paths should have equal `ProjectPanelOrdMatch`, so we should
+        // not have any duplicates after building the final list.
+        for new_match in new_history_matches
+            .into_values()
+            .chain(new_search_matches.into_iter())
+        {
+            match self.position(&new_match) {
+                Ok(_duplicate) => continue,
+                Err(i) => {
+                    self.matches.insert(i, new_match);
+                    if self.matches.len() == 100 {
+                        break;
                     }
-                    (Match::Search(match_a), Match::Search(match_b)) => match_b.cmp(match_a),
                 }
-                .then(history_score_a.cmp(history_score_b))
-            })
-            .take(100)
-            .map(|(_, m)| m)
-            .collect();
+            }
+        }
+    }
+
+    /// If a < b, then a is a worse match, aligning with the `ProjectPanelOrdMatch` ordering.
+    fn cmp_matches(
+        separate_history: bool,
+        currently_opened: Option<&FoundPath>,
+        a: &Match,
+        b: &Match,
+    ) -> cmp::Ordering {
+        assert!(a.panel_match().is_some() && b.panel_match().is_some());
+
+        match (&a, &b) {
+            // bubble currently opened files to the top
+            (Match::History { path, .. }, _) if Some(path) == currently_opened => {
+                cmp::Ordering::Greater
+            }
+            (_, Match::History { path, .. }) if Some(path) == currently_opened => {
+                cmp::Ordering::Less
+            }
+
+            (Match::History { .. }, Match::Search(_)) if separate_history => cmp::Ordering::Greater,
+            (Match::Search(_), Match::History { .. }) if separate_history => cmp::Ordering::Less,
+
+            _ => a.panel_match().cmp(&b.panel_match()),
+        }
     }
 }
 
-fn matching_history_item_paths<'a>(
+/// Produces a HashMap of history entries matching the query.
+/// `currently_opened` also treated as a history entry.
+/// HashMap is used because it's convenient for eliminating duplicate
```

Comment:
Here and below comes something quite implementation-specific, let's remove it before it gets obsolete.